src/platform/OpenThread/GenericNetworkCommissioningThreadDriver.cpp

```diff
@@ -86,23 +84,15 @@ CHIP_ERROR GenericThreadDriver::RevertConfiguration()
     // If no backup could be found, it means that the network configuration has not been modified
     // since the fail-safe was armed, so return with no error.
     ReturnErrorCodeIf(error == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND, CHIP_NO_ERROR);
-    ReturnErrorOnFailure(error);
-
-    ChipLogError(NetworkProvisioning, "Found Thread configuration backup: reverting configuration");
 
-    // Not all KVS implementations support zero-length values, so handle a special value representing an empty dataset.
-    ByteSpan dataset(datasetBytes, datasetLength);
-
-    if (dataset.data_equal(ByteSpan(kEmptyDataset)))
-    {
-        dataset = {};
-    }
+    // Always remove the backup, regardless if it can be successfully restored.
+    KeyValueStoreMgr().Delete(key.FailSafeNetworkConfig());
+    ReturnErrorOnFailure(error);
 
-    ReturnErrorOnFailure(mStagingNetwork.Init(dataset));
+    ReturnErrorOnFailure(mStagingNetwork.Init(ByteSpan(datasetBytes, datasetLength)));
```

Comment:
If we reboot again between the delete and the `mStagingNetwork.Init` call, won't that lose our fail-safe recovery.

It seems to me like the delete should still be the last thing here, but should be done even if the previous steps failed, as long as we _tried_ those steps.

Reply:
Good point. Reworked according to your suggestion.